node/network/collator-protocol/src/validator_side/peer_slots.rs

```diff
@@ -0,0 +1,1046 @@
+#!`[allow(dead_code)]
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{CollatorId, CollatorProtocolMessage, PeerId, SubsystemContext, LOG_TARGET};
+use std::time::{Duration, Instant};
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+use std::ops::{Index, IndexMut};
+
+use lru::LruCache;
+use rand::{thread_rng, Rng};
+
+use sp_keystore::SyncCryptoStorePtr;
+
+use polkadot_node_network_protocol::{UnifiedReputationChange as Rep, View};
+use polkadot_primitives::v1::{Hash, Id as ParaId};
+use polkadot_subsystem::{messages::{AllMessages, NetworkBridgeMessage}, SubsystemSender};
+
+// CACHE_SIZE determines the size (in number of entries) of in-memory cache maintained for the
+// purposes of determining a collator's fitness. Metrics for collators that do not fit in the
+// Cache are stored in a database.
+const CACHE_SIZE: usize = 1_000;
+// RESERVOIR SIZE determines the upper bound on the number of collator connections
+#[cfg(not(test))]
+pub(super) const RESERVOIR_SIZE: usize = 128;
```

Comment:
this should be hooked into https://github.com/paritytech/polkadot/blob/8be63cd653e6ee73e320a3aaa6cfacb5f13b5c57/node/network/protocol/src/peer_set.rs#L75

alternatively, we can make the latter limit much higher